File: src/show_me_tokens/adapters/qoderwork.py
"""Adapter for QoderWork local agents database."""

from __future__ import annotations

import json
import re
import sqlite3
from datetime import datetime, timezone
from pathlib import Path
from typing import Any, ClassVar

from show_me_tokens.adapters.base import AgentAdapter
from show_me_tokens.models import SessionFilters, SessionUsage, TokenSummary

_TOKEN_KEY_RE = re.compile(r"(^|_)(prompt|input|completion|output|generated|total|cached)_?tokens?$", re.IGNORECASE)


class QoderWorkAdapter(AgentAdapter):
    """Reads QoderWork ``data/agents.db``."""

    name: ClassVar[str] = "qoderwork"

    def default_db_path(self) -> Path:
        return (
            Path.home()
            / "Library"
            / "Application Support"
            / "QoderWork"
            / "data"
            / "agents.db"
        )

    def collect_sessions(
        self,
        db_path: Path,
        filters: SessionFilters,
    ) -> list[SessionUsage]:
        conn = self._connect_readonly(db_path)
        try:
            rows = conn.execute(
                """
                SELECT
                    sc.id AS sub_chat_id,
                    sc.name AS sub_chat_name,
                    sc.chat_id,
                    sc.session_id,
                    sc.mode,
                    sc.model_level,
                    sc.created_at AS sub_chat_created_at,
                    sc.updated_at AS sub_chat_updated_at,
                    c.name AS chat_name,
                    c.worktree_path,
                    c.chat_type,
                    p.name AS project_name,
                    p.path AS project_path
                FROM sub_chats sc
                JOIN chats c ON c.id = sc.chat_id
                JOIN projects p ON p.id = c.project_id
                ORDER BY COALESCE(sc.updated_at, sc.created_at, 0) DESC
                """
            ).fetchall()

            sessions: list[SessionUsage] = []
            for row in rows:
                usage = self._build_session(conn, dict(row), filters)
                if usage is not None:
                    sessions.append(usage)
            return sessions
        finally:
            conn.close()

    def _connect_readonly(self, db_path: Path) -> sqlite3.Connection:
        if not db_path.exists():
            raise FileNotFoundError(f"QoderWork database not found: {db_path}")
        uri = f"file:{db_path}?mode=ro"
        conn = sqlite3.connect(uri, uri=True)
        conn.row_factory = sqlite3.Row
        return conn

    def _build_session(
        self,
        conn: sqlite3.Connection,
        row: dict[str, Any],
        filters: SessionFilters,
    ) -> SessionUsage | None:
        sub_chat_id = row["sub_chat_id"]
        session_id = row["session_id"] or sub_chat_id

        if filters.session_id is not None:
            if filters.session_id not in (sub_chat_id, session_id):
                return None

        workspace_path = self._normalize_path(
            row.get("worktree_path") or row.get("project_path")
        )
        if filters.workspace is not None:
            if workspace_path is None:
                return None
            try:
                if Path(workspace_path).resolve() != filters.workspace.resolve():
                    return None
            except OSError:
                return None

        created_at = self._epoch_to_datetime(row.get("sub_chat_created_at"))
        updated_at = self._epoch_to_datetime(
            row.get("sub_chat_updated_at") or row.get("sub_chat_created_at")
        )

        if not self._in_date_range(updated_at, filters):
            return None

        messages = conn.execute(
            """
            SELECT sequence, role, parts, metadata
            FROM messages
            WHERE sub_chat_id = ?
            ORDER BY sequence ASC
            """,
            (sub_chat_id,),
        ).fetchall()

        token_summary = TokenSummary()
        model_name: str | None = None
        event_count = 0

        for msg in messages:
            parts = self._try_json(msg["parts"])
            metadata = self._try_json(msg["metadata"])
            events: list[dict[str, Any]] = []
            self._collect_token_events(parts, events)
            self._collect_token_events(metadata, events)

            for event in events:
                input_tokens = self._numeric(
                    event.get("prompt_tokens")
                    or event.get("input_tokens")
                    or event.get("promptTokens")
                    or event.get("inputTokens")
                )
                output_tokens = self._numeric(
                    event.get("completion_tokens")
                    or event.get("output_tokens")
                    or event.get("generated_tokens")
                    or event.get("completionTokens")
                    or event.get("outputTokens")
                    or event.get("generatedTokens")
                )
                total_tokens = self._numeric(
                    event.get("total_tokens") or event.get("totalTokens")
                )
                cached_tokens = self._numeric(
                    event.get("cached_tokens") or event.get("cachedTokens")
                )

                token_summary.input_tokens += input_tokens
                token_summary.output_tokens += output_tokens
                if total_tokens:
                    token_summary.total_tokens = (
                        token_summary.total_tokens or 0
                    ) + total_tokens
                if cached_tokens:
                    token_summary.cache_read_input_tokens = (
                        token_summary.cache_read_input_tokens or 0
                    ) + cached_tokens
                event_count += 1

            if model_name is None:
                for source in (metadata, parts):
                    name = self._extract_model_name(source)
                    if name:
                        model_name = name
                        break

        token_summary.recompute_total()

        if model_name is None:
            model_name = row.get("model_level")

        note: str | None = None
        if event_count == 0:
            note = (
                "Token usage unavailable: no comparable token fields found "
                "in local agents.db"
            )
            token_summary.total_tokens = None

        return SessionUsage(
            agent=self.name,
            session_id=session_id,
            chat_id=row.get("chat_id"),
            project_name=row.get("project_name"),
            workspace_path=workspace_path,
            model_name=model_name,
            created_at=created_at,
            updated_at=updated_at,
            tokens=token_summary,
            note=note,
            raw_source={"sub_chat_id": sub_chat_id, "mode": row.get("mode")},
        )

    @staticmethod
    def _collect_token_events(value: Any, events: list[dict[str, Any]]) -> None:
        if value is None:
            return
        if isinstance(value, list):
            for item in value:
                QoderWorkAdapter._collect_token_events(item, events)
            return
        if not isinstance(value, dict):
            return

        keys = value.keys()
        if any(_TOKEN_KEY_RE.search(key) for key in keys):
            events.append(value)

        for nested in value.values():
            QoderWorkAdapter._collect_token_events(nested, events)

    @staticmethod
    def _normalize_path(value: str | None) -> str | None:
        if not value:
            return None
        try:
            return str(Path(value).resolve())
        except (OSError, ValueError):
            return value

    @staticmethod
    def _epoch_to_datetime(value: int | None) -> datetime | None:
        if value is None:
            return None
        try:
            number = int(value)
            if not number:
                return None
            ms = number if number > 100000000000 else number * 1000
            return datetime.fromtimestamp(ms / 1000, tz=timezone.utc)
        except (OSError, ValueError, OverflowError):
            return None

    @staticmethod
    def _in_date_range(value: datetime | None, filters: SessionFilters) -> bool:
        if value is None:
            return True
        if filters.since is not None and value.date() < filters.since.date():
            return False
        if filters.until is not None and value.date() > filters.until.date():
            return False
        return True

    @staticmethod
    def _try_json(text: str | None) -> Any:
        if not text:
            return None
        try:
            return json.loads(text)
        except Exception:
            return None

    @staticmethod
    def _numeric(value: Any) -> int:
        if value is None:
            return 0
        try:
            number = int(value)
            return number if number >= 0 else 0
        except (TypeError, ValueError):
            return 0

    @staticmethod
    def _extract_model_name(data: Any) -> str | None:
        if not isinstance(data, dict):
            return None
        for key in ("model_name", "model", "modelId", "model_id", "modelLevel"):
            value = data.get(key)
            if isinstance(value, str) and value.strip():
                return value.strip()
        return None

File: src/adapters/qoderwork.ts
import Database from 'better-sqlite3';
import fs from 'node:fs';
import path from 'node:path';
import os from 'node:os';
import { AgentAdapter } from './base.js';
import { SessionFilters, SessionUsage } from '../shared/types.js';
import { createTokenSummary, recomputeTotal } from '../models/index.js';

const TOKEN_KEY_RE = /(^|_)(prompt|input|completion|output|generated|total|cached)_?tokens?$/i;

export class QoderWorkAdapter extends AgentAdapter {
  readonly name = 'qoderwork';

  defaultDbPath(): string {
    return path.join(
      os.homedir(),
      'Library',
      'Application Support',
      'QoderWork',
      'data',
      'agents.db'
    );
  }

  collectSessions(dbPath: string, filters: SessionFilters): SessionUsage[] {
    if (!this._fileExists(dbPath)) {
      throw new Error(`QoderWork database not found: ${dbPath}`);
    }
    const db = new Database(dbPath, { readonly: true });
    try {
      const rows = db
        .prepare(
          `
          SELECT
            sc.id AS sub_chat_id,
            sc.name AS sub_chat_name,
            sc.chat_id,
            sc.session_id,
            sc.mode,
            sc.model_level,
            sc.ext,
            sc.created_at AS sub_chat_created_at,
            sc.updated_at AS sub_chat_updated_at,
            c.name AS chat_name,
            c.worktree_path,
            c.chat_type,
            p.name AS project_name,
            p.path AS project_path
          FROM sub_chats sc
          JOIN chats c ON c.id = sc.chat_id
          JOIN projects p ON p.id = c.project_id
          ORDER BY COALESCE(sc.updated_at, sc.created_at, 0) DESC
          `
        )
        .all() as Array<{
          sub_chat_id: string;
          sub_chat_name: string | null;
          chat_id: string;
          session_id: string | null;
          mode: string | null;
          model_level: string | null;
          ext: string | null;
          sub_chat_created_at: number | null;
          sub_chat_updated_at: number | null;
          chat_name: string | null;
          worktree_path: string | null;
          chat_type: string | null;
          project_name: string | null;
          project_path: string | null;
        }>;

      const sessions: SessionUsage[] = [];
      for (const row of rows) {
        const usage = this._buildSession(db, row, filters);
        if (usage !== null) {
          sessions.push(usage);
        }
      }
      return sessions;
    } finally {
      db.close();
    }
  }

  private _buildSession(
    db: Database.Database,
    row: {
      sub_chat_id: string;
      sub_chat_name: string | null;
      chat_id: string;
      session_id: string | null;
      mode: string | null;
      model_level: string | null;
      ext: string | null;
      sub_chat_created_at: number | null;
      sub_chat_updated_at: number | null;
      chat_name: string | null;
      worktree_path: string | null;
      chat_type: string | null;
      project_name: string | null;
      project_path: string | null;
    },
    filters: SessionFilters
  ): SessionUsage | null {
    const subChatId = row.sub_chat_id;
    const sessionId = row.session_id ?? subChatId;

    if (filters.sessionId !== undefined) {
      if (filters.sessionId !== subChatId && filters.sessionId !== sessionId) {
        return null;
      }
    }

    const workspacePath = this._normalizePath(row.worktree_path ?? row.project_path);
    if (filters.workspace !== undefined) {
      if (workspacePath === null) {
        return null;
      }
      try {
        if (path.resolve(workspacePath) !== path.resolve(filters.workspace)) {
          return null;
        }
      } catch {
        return null;
      }
    }

    const createdAt = this._epochToIso(row.sub_chat_created_at);
    const updatedAt = this._epochToIso(
      row.sub_chat_updated_at ?? row.sub_chat_created_at
    );

    if (!this._inDateRange(updatedAt, filters)) {
      return null;
    }

    const title =
      (row.sub_chat_name || row.chat_name) ?? subChatId;

    const ext = this._tryJson(row.ext);
    let modelName = this._extractExtModel(ext) ?? row.model_level;

    const messages = db
      .prepare(
        `
        SELECT sequence, role, parts, metadata
        FROM messages
        WHERE sub_chat_id = ?
        ORDER BY sequence ASC
        `
      )
      .all(subChatId) as Array<{
        sequence: number;
        role: string;
        parts: string | null;
        metadata: string | null;
      }>;

    const tokenSummary = createTokenSummary();
    let eventCount = 0;

    for (const msg of messages) {
      const parts = this._tryJson(msg.parts);
      const metadata = this._tryJson(msg.metadata);
      const events: Array<Record<string, unknown>> = [];
      this._collectTokenEvents(parts, events);
      this._collectTokenEvents(metadata, events);

      for (const event of events) {
        const inputTokens =
          this._numeric(
            event.prompt_tokens ??
              event.input_tokens ??
              event.promptTokens ??
              event.inputTokens
          );
        const outputTokens =
          this._numeric(
            event.completion_tokens ??
              event.output_tokens ??
              event.generated_tokens ??
              event.completionTokens ??
              event.outputTokens ??
              event.generatedTokens
          );
        const totalTokens = this._numeric(event.total_tokens ?? event.totalTokens);
        const cachedTokens = this._numeric(
          event.cached_tokens ?? event.cachedTokens
        );

        tokenSummary.inputTokens += inputTokens;
        tokenSummary.outputTokens += outputTokens;
        if (totalTokens) {
          tokenSummary.totalTokens = (tokenSummary.totalTokens ?? 0) + totalTokens;
        }
        if (cachedTokens) {
          tokenSummary.cacheReadInputTokens =
            (tokenSummary.cacheReadInputTokens ?? 0) + cachedTokens;
        }
        eventCount += 1;
      }

      if (modelName === null || modelName === undefined) {
        for (const source of [metadata, parts]) {
          const name = this._extractModelName(source);
          if (name) {
            modelName = name;
            break;
          }
        }
      }
    }

    recomputeTotal(tokenSummary);

    if (filters.modelName !== undefined) {
      const modelNameStr = typeof modelName === 'string' ? modelName : '';
      if (!modelNameStr.toLowerCase().includes(filters.modelName.toLowerCase())) {
        return null;
      }
    }

    let note: string | null = null;
    if (eventCount === 0) {
      note =
        'Token usage unavailable: no comparable token fields found in local agents.db';
      tokenSummary.totalTokens = null;
    }

    return {
      agent: this.name,
      sessionId,
      title,
      chatId: row.chat_id,
      projectName: row.project_name,
      workspacePath,
      modelName: modelName ?? null,
      createdAt,
      updatedAt,
      tokens: tokenSummary,
      note,
      rawSource: { sub_chat_id: subChatId, mode: row.mode },
    };
  }

  private _fileExists(dbPath: string): boolean {
    try {
      return fs.statSync(dbPath).isFile();
    } catch {
      return false;
    }
  }

  private _extractExtModel(data: unknown): string | null {
    if (!this._isRecord(data)) {
      return null;
    }
    const snapshot = data.contextUsageSnapshot;
    if (!this._isRecord(snapshot)) {
      return null;
    }
    const value = snapshot.model;
    if (typeof value === 'string' && value.trim()) {
      return value.trim();
    }
    return null;
  }

  private _collectTokenEvents(value: unknown, events: Array<Record<string, unknown>>): void {
    if (value === null || value === undefined) {
      return;
    }
    if (Array.isArray(value)) {
      for (const item of value) {
        this._collectTokenEvents(item, events);
      }
      return;
    }
    if (!this._isRecord(value)) {
      return;
    }

    const keys = Object.keys(value);
    if (keys.some((key) => TOKEN_KEY_RE.test(key))) {
      events.push(value);
    }

    for (const nested of Object.values(value)) {
      this._collectTokenEvents(nested, events);
    }
  }

  private _normalizePath(value: string | null): string | null {
    if (!value) {
      return null;
    }
    try {
      return path.resolve(value);
    } catch {
      return value;
    }
  }

  private _epochToIso(value: number | null): string | null {
    if (value === null || value === undefined) {
      return null;
    }
    try {
      let ms = Number(value);
      if (Number.isNaN(ms)) {
        return null;
      }
      if (ms < 100000000000) {
        ms *= 1000;
      }
      return new Date(ms).toISOString();
    } catch {
      return null;
    }
  }

  private _inDateRange(value: string | null, filters: SessionFilters): boolean {
    if (value === null) {
      return true;
    }
    const date = new Date(value);
    if (Number.isNaN(date.getTime())) {
      return true;
    }
    if (filters.since !== undefined) {
      const since = new Date(filters.since);
      if (date.getTime() < since.getTime()) {
        return false;
      }
    }
    if (filters.until !== undefined) {
      const until = new Date(filters.until);
      until.setDate(until.getDate() + 1);
      if (date.getTime() >= until.getTime()) {
        return false;
      }
    }
    return true;
  }

  private _tryJson(text: string | null): unknown {
    if (!text) {
      return null;
    }
    try {
      return JSON.parse(text);
    } catch {
      return null;
    }
  }

  private _extractModelName(data: unknown): string | null {
    if (!this._isRecord(data)) {
      return null;
    }
    const keys = [
      'model_name',
      'model',
      'modelId',
      'model_id',
      'modelLevel',
    ];
    for (const key of keys) {
      const value = data[key];
      if (typeof value === 'string' && value.trim()) {
        return value.trim();
      }
    }
    return null;
  }

  private _isRecord(value: unknown): value is Record<string, unknown> {
    return typeof value === 'object' && value !== null && !Array.isArray(value);
  }

  private _numeric(value: unknown): number {
    if (value === null || value === undefined) {
      return 0;
    }
    try {
      const num = typeof value === 'number' ? value : Number(value);
      return Number.isFinite(num) && num >= 0 ? Math.floor(num) : 0;
    } catch {
      return 0;
    }
  }
}
